sfge/src/main/java/com/salesforce/rules/dmlinloop/DmlInLoopVisitor.java

```diff
@@ -0,0 +1,130 @@
+package com.salesforce.rules.dmlinloop;
+
+import com.salesforce.graph.symbols.SymbolProvider;
+import com.salesforce.graph.vertex.*;
+import com.salesforce.rules.DmlUtil;
+import com.salesforce.rules.Violation;
+import com.salesforce.rules.ops.visitor.LoopDetectionVisitor;
+import java.util.HashSet;
+import java.util.Optional;
+import java.util.Set;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public class DmlInLoopVisitor extends LoopDetectionVisitor {
+
+    private static final Logger LOGGER = LogManager.getLogger(DmlInLoopVisitor.class);
+
+    /** Represents the path entry point that this visitor is walking */
+    private final SFVertex sourceVertex;
+
+    /** Represents the DML statement that is possibly inside a loop */
+    private final BaseSFVertex sinkVertex;
+
+    /** Collects violation information */
+    private final HashSet<Violation.PathBasedRuleViolation> violations;
+
+    // DmlStatementVertex represents statements like insert a;
+    DmlInLoopVisitor(SFVertex sourceVertex, DmlStatementVertex sinkVertex) {
+        this.sourceVertex = sourceVertex;
+        this.sinkVertex = sinkVertex;
+        this.violations = new HashSet<>();
+    }
+
+    // MethodCallExpressionVertex is necessary to detect Database.<whatever> methods
+    DmlInLoopVisitor(SFVertex sourceVertex, MethodCallExpressionVertex sinkVertex) {
+        this.sourceVertex = sourceVertex;
+        this.sinkVertex = sinkVertex;
+        this.violations = new HashSet<>();
+    }
+
+    // DmlStatementVertex represents statements like [SELECT Id, Age FROM Accounts]
+    DmlInLoopVisitor(SFVertex sourceVertex, SoqlExpressionVertex sinkVertex) {
+        this.sourceVertex = sourceVertex;
+        this.sinkVertex = sinkVertex;
+        this.violations = new HashSet<>();
+    }
```

Comment:
Can these be combined into a single constructor that accepts a `BaseSFVertex` and throws a `ProgrammingException` if that vertex is of an unacceptable type?
Alternatively, can the comments above each constructor be turned into JDoc-style comments?